# Synthetic_virome
TLDR: synthesize illumina reads for soil samples to test viral identification tool 
This script is designed to take an input viral multifasta and random RefSoil+ bacterial sequences and generate illumina reads simulating varying concentrations of viral nucleotides in the sample.
*hypotheses to test
	What viral identification tools work best?
	What are the effects of lower coverages?
	What are the effects of more conplex background (more bacteria, fungus, and non-target virus)?
